include/clap/preset-discovery.h

```diff
@@ -0,0 +1,232 @@
+/**
+   Preset Discovery API.
+
+   Preset Discovery enables a plug-in host to identify where presets are found, what
+   extensions they have, which plug-ins they apply to, and other metadata associated with the
+   presets so that they can be indexed and searched for quickly within the plug-in host's browser.
+
+   This has a number of advantages for the user:
+   - it allows them to browse for presets from one central location in a consistent way
+   - the user can browse for presets without having to commit to a particular plug-in first
+
+   The API works as follow to index presets and presets metadata:
+   1. get the preset discovery factory
+   2. create a preset provider
+   3. list locations and file extensions from the preset provider
+   4. monitor file system changes on the given locations
+   5. read metadata for each preset files
+   6. fetch information about collections from the preset provider
+
+   Then to load a preset, use ext/draft/preset-load.h
+
+   The design of this API deliberately does not define a fixed set tags or categories. It is the
+   plug-in host's job to try to intelligently map the raw list of features that are found for a
+   preset and to process this list to generate something that makes sense for the host's tagging and
+   categorization system. The reason for this is to reduce the work for a plug-in developer to add
+   Preset Discovery support for their existing preset file format and not have to be concerned with
+   all the different hosts and how they want to receive the metadata.
+*/
+
+#pragma once
+
+#include "private/std.h"
+#include "private/macros.h"
+#include "string-sizes.h"
+#include "version.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** Receiver that receives the metadata for a single preset file.
+ * The host would define the various callbacks in this interface and the preset parser function
+ * would then call them.
+ *
+ * This interface isn't thread-safe.
+ */
+typedef const struct clap_preset_metadata_receiver {
+   /** If there is an error reading metadata from a file this should be called with an error
+    * message. */
+   void(CLAP_ABI *on_error)(const struct clap_preset_metadata_receiver *receiver,
+                            const char                                 *error_message);
+
```

Comment:
Please add an error code here, e.g. "errno" based.
A program (and user) can handle different error scenarios that way, e.g.

- EPERM: ask root/admin to give access rights
- ENOSPC: ask the user to store on a different device (upon saving)
- ESTALE: broken fuse/NFS handle, user needs to remount
- EIO: user needs to fix disk or USB stick, etc
- EILSEQ/EFTYPE: content/file type wrong
- ENOTSUP/ENOSYS: loading not implemented

Reply:
Added!